src/coreclr/jit/helperexpansion.cpp

```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+    // We cannot morph stmt yet as we may modify it further below, and the morphing
+    // could invalidate callUse
+    while ((newFirstStmt != nullptr) && (newFirstStmt != stmt))
+    {
+        fgMorphStmtBlockOps(block, newFirstStmt);
+        newFirstStmt = newFirstStmt->GetNextStmt();
+    }
+
+    // We don't need this flag anymore.
+    call->gtCallMoreFlags &= ~GTF_CALL_M_SPECIAL_INTRINSIC;
+
+    GenTreeLclVar* resultLcl = nullptr;
+
+    // Grab a temp to store the result.
+    // The result corresponds the number of bytes written to dstPtr (int32).
+    assert(call->TypeIs(TYP_INT));
+    const unsigned resultLclNum   = lvaGrabTemp(true DEBUGARG("local for result"));
+    lvaTable[resultLclNum].lvType = TYP_INT;
+    resultLcl                     = gtNewLclvNode(resultLclNum, TYP_INT);
+    *callUse                      = resultLcl;
+    fgMorphStmtBlockOps(block, stmt);
+    gtUpdateStmtSideEffects(stmt);
+
+    // srcLenCns is the length of the string literal in chars (UTF16)
+    // but we're going to use the same value as the "bytesWritten" result in the fast path and in the length check.
+    GenTree* srcLenCnsNode = gtNewIconNode(srcLenCns);
+    fgUpdateConstTreeValueNumber(srcLenCnsNode);
+
+    // We're going to insert the following blocks:
+    //
+    //  prevBb:
+    //
+    //  lengthCheckBb:
+    //      <side-effects>
+    //      bytesWritten = -1;
+    //      if (dstLen <srcLen)
+    //          goto block;
+    //
+    //  fastpathBb:
+    //      bytesWritten = unrolled copy;
+    //
+    //  block:
+    //      use(bytesWritten)
+    //
+
+    // or in case if noSizeCheck is true:
+
+    //  prevBb:
+    //
+    //  lengthCheckBb:
+    //      <side-effects>
+    //
+    //  fastpathBb:
+    //      bytesWritten = unrolled copy;
+    //
+    //  block:
+    //      use(bytesWritten)
+    //
+
+    //
+    // Block 1: lengthCheckBb (we check that dstLen < srcLen)
+    //  In case if destIsKnownToFit is true we'll use this block to keep side-effects of the original arguments.
+    //  and it will be a fall-through block.
+    //
+    BasicBlock* lengthCheckBb = fgNewBBafter(noSizeCheck ? BBJ_NONE : BBJ_COND, prevBb, true);
+    lengthCheckBb->bbFlags |= BBF_INTERNAL;
+
+    // In 99% cases "this" is expected to be "static readonly UTF8EncodingSealed s_default"
+    // which is a static readonly object that is never null
+    const bool thisIsKnownNonNull =
+        thisArg->gtVNPair.BothEqual() && vnStore->IsKnownNonNull(thisArg->gtVNPair.GetLiberal());
+
+    // Spill all original arguments to locals in the lengthCheckBb to preserve all possible side-effects.
+    thisArg = SpillExpression(this, thisArg, lengthCheckBb, debugInfo);
+    srcPtr  = SpillExpression(this, srcPtr, lengthCheckBb, debugInfo);
+    srcLen  = SpillExpression(this, srcLen, lengthCheckBb, debugInfo);
+    dstPtr  = SpillExpression(this, dstPtr, lengthCheckBb, debugInfo);
+    dstLen  = SpillExpression(this, dstLen, lengthCheckBb, debugInfo);
```

Comment:
Is this necessary? These should have been spilled by `gtSplitTree`.

Reply:
These node pointers are from before `gtSplitTree` made changes, it seems very dangerous to rely on them. You should rather reload them from the individual `CallArg`s.